pir/cpp/database.cpp

```diff
@@ -44,90 +48,132 @@ StatusOr<std::shared_ptr<PIRDatabase>> PIRDatabase::Create(
+  }
 
-      if (relin_keys != nullptr) {
-        evaluator.relinearize_inplace(temp_ct, *relin_keys);
-        if (decryptor != nullptr) {
-          std::cout << depth_string << "i = " << i << " noise budget relin "
-                    << decryptor->invariant_noise_budget(temp_ct) << std::endl;
+ private:
+  /**
+   * Recursive function to do the dot product of each dimension with the db.
+   * Calls itself to move down dimensions until you get to the bottom dimension.
+   * Bottom dimension just does a dot product with the DB, and returns the
+   * result. Upper levels then take those results, and dot product again with
+   * the selection vector, until you get back to the top. NB: Database iterator
+   * is kept at the class level so that we move through the database one element
+   * at a time.
+   *
+   * @param[in] dimensions List of remaining demainsion sizes.
+   * @param[in] selection_vector_it Iterator into the start of the selection
+   *  vector for the current depth.
+   * @param[in] depth Current depth.
+   */
+  Ciphertext multiply(vector<uint32_t> dimensions,
+                      vector<Ciphertext>::const_iterator selection_vector_it,
+                      size_t depth) {
+    const size_t this_dimension = dimensions[0];
+    auto remaining_dimensions =
+        vector<uint32_t>(dimensions.begin() + 1, dimensions.end());
+
+    string depth_string(depth, ' ');
+
+    Ciphertext result;
+    bool first_pass = true;
+    for (size_t i = 0; i < this_dimension; ++i) {
+      // make sure we don't go past end of DB
+      if (database_it_ == database_.end()) break;
+      Ciphertext temp_ct;
+      if (remaining_dimensions.empty()) {
+        // base case: have to multiply against DB
+        evaluator_.multiply_plain(*(selection_vector_it + i), *(database_it_++),
+                                  temp_ct);
+        print_noise(depth, i, "base", temp_ct);
+
+      } else {
+        temp_ct = multiply(remaining_dimensions,
+                           selection_vector_it + this_dimension, depth + 1);
+        print_noise(depth, i, "recurse", temp_ct);
+
+        evaluator_.multiply_inplace(temp_ct, *(selection_vector_it + i));
+        print_noise(depth, i, "mult", temp_ct);
+
+        if (relin_keys_ != nullptr) {
+          evaluator_.relinearize_inplace(temp_ct, *relin_keys_);
+          print_noise(depth, i, "relin", temp_ct);
         }
       }
-    }
 
-    if (first_pass) {
-      result = temp_ct;
-      first_pass = false;
-    } else {
-      evaluator.add_inplace(result, temp_ct);
-      if (decryptor != nullptr) {
-        std::cout << depth_string << "i = " << i << " noise budget result "
-                  << decryptor->invariant_noise_budget(result) << std::endl;
+      if (first_pass) {
+        result = temp_ct;
+        first_pass = false;
+      } else {
+        evaluator_.add_inplace(result, temp_ct);
+        print_noise(depth, i, "result", temp_ct);
       }
     }
+    if (decryptor_ != nullptr) {
```

Comment:
Why not print_noise?

Reply:
print_noise needs a variable for the iteration count. This is the only case that doesn't use that, and I couldn't think of a clean way of making print_noise do both.

Reply:
Maybe we can adapt print_noise. 
Or add an overloaded method without the iteration count, or make the count optional in the existing one.

It would be easier the read the multiply method without the debug code.

Reply:
I was thinking about it, and you're right it's more readable if we adapt print_noise. Then I realized that if we use std::optional, that makes things easy. I've done so, and I think it's cleaner. Thanks for this!